pointcloud/src/main/scala/geotrellis/pointcloud/pipeline/PipelineExpressions.scala

```diff
@@ -0,0 +1,487 @@
+  approximate: Option[Boolean] = None,
+  `type`: FilterType = pmf
+) extends PipelineExpr
+
+case class PoissonFilter(
+  depth: Option[Int] = None,
+  pointWeight: Option[Double] = None,
+  `type`: FilterType = poisson
+) extends PipelineExpr
+
+case class PredicateFilter(
+  script: String,
+  module: String,
+  function: String,
+  `type`: FilterType = predicate
+) extends PipelineExpr
+
+case class ProgrammableFilter(
+  script: String,
+  module: String,
+  function: String,
+  source: String,
+  addDimenstion: Option[String] = None,
+  `type`: FilterType = programmable
+) extends PipelineExpr
+
+case class RadialDensityFilter(
+  radius: Option[Double] = None,
+  `type`: FilterType = radialdensity
+) extends PipelineExpr
+
+case class RandomizeFilter(
+  `type`: FilterType = randomize
+) extends PipelineExpr
+
+case class RangeFilter(
+  limits: Option[String] = None,
+  `type`: FilterType = range
+) extends PipelineExpr
+
+case class ReprojectionFilter(
+  outSrs: String,
+  inSrs: Option[String] = None,
+  tag: Option[String] = None,
+  `type`: FilterType = reprojection
+) extends PipelineExpr
+
+case class SampleFilter(
+  radius: Option[Double] = None,
+  `type`: FilterType = sample
+) extends PipelineExpr
+
+case class SmrfFilter(
+  cell: Option[Double] = None,
+  classify: Option[Boolean] = None,
+  cut: Option[Double] = None,
+  extract: Option[Boolean] = None,
+  slope: Option[Double] = None,
+  threshold: Option[Double] = None,
+  window: Option[Double] = None,
+  `type`: FilterType = smrf
+) extends PipelineExpr
+
+case class SortFilter(
+  dimension: String,
+  `type`: FilterType = sort
+) extends PipelineExpr
+
+case class SplitterFilter(
+  length: Option[Int] = None,
+  originX: Option[Double] = None,
+  originY: Option[Double] = None,
+  `type`: FilterType = splitter
+) extends PipelineExpr
+
+case class StatsFilter(
+  dimenstions: Option[String] = None,
+  enumerate: Option[String] = None,
+  count: Option[Int] = None,
+  `type`: FilterType = stats
+) extends PipelineExpr
+
+case class TransformationFilter(
+  matrix: String,
+  `type`: FilterType = transformation
+) extends PipelineExpr
+
+case class VoxelGridFilter(
+  leafX: Option[Double] = None,
+  leafY: Option[Double] = None,
+  leafZ: Option[Double] = None,
+  `type`: FilterType = voxelgrid
+) extends PipelineExpr
+
+case class PipelineConstructor(list: List[PipelineExpr]) {
+  def ~(e: PipelineExpr): PipelineConstructor = PipelineConstructor(list :+ e)
+  def map[B](f: PipelineExpr => B): List[B] = list.map(f)
+  def mapExpr(f: PipelineExpr => PipelineExpr): PipelineConstructor = PipelineConstructor(list.map(f))
+  def tail: List[PipelineExpr] = list.tail
+  def head: PipelineExpr = list.head
+  def json: Json =
+    Json.obj(
+      "pipeline" -> list
+        .map(
+          _.asJsonObject
+            .remove("class_type") // remove type
+            .filter { case (key, value) => !value.isNull } // cleanup options
+        ).asJson
+    )
+}
```

Comment:
It might be possible to avoid this class altogether.

```scala
 /* "Lightest" way */
 sealed trait PipelineExpr

 object Pipeline {
   type Pipeline = List[PipelineExpr]

   def json(list: Pipeline): Json = ... // same definition you have                                       
 }

 // Define a pipeline like:                                                                               
 val pipeline: Pipeline = List(
   foo,
   bar,
   baz
 )

 // Get the JSON                                                                                          
 val json: Json = Pipline.json(pipeline)
```
or
```scala
 /* Fancy way */
 sealed trait PipelineExpr {
   /* The next stage in the pipeline */
   def next: Option[PipelineExpr]

   def json: Json = ???
 }

 object PipelineExpr {
   val pipelineSemi: Semigroup[PipelineExpr] = new Semigroup[PipelineExpr] { ... }
 }

 // Define a pipeline like:                                                                               
 val json: Json = List(
   foo,
   bar,
   baz
 ).reduce(_ mappend _).json

// or like this
val json: Json = (foo |+| bar |+| baz).json
```